services/regulated-assets-approval-server: update README and refactor tx_approve.go file (#3642)
stellar/go: services/regulated-assets-approval-server/README.md

# regulated-assets-approval-server

```sh

> Status: supporting SEP-8 transactions revision with a simplified rule:
> - only revises payment type operations
> - payments with amount lower than or equal the configured threshold are revised successfully
> - payments with amount exceeding the threshold need further action

Reviewer: 🔠 It was a little confusing understanding what `revised successfully` means. I think it is worth talking about when txns are accepted or not, and separately when they are revised. If you feel this misrepresents what the service does, please disregard.
```suggestion
Status: supporting SEP-8 transactions revision with a simplified rule:
- only accepts payment operations
- payments with amount lower than or equal the configured threshold are accepted without requiring further action
- payments with amount exceeding the threshold require further action
- payments without the auth/revoke sandwich are revised to include them
```

Author: I agree the concept of accepting and revising a transaction are currently intertwined, that's because the current version of the service doesn't implement the SEP-8 "success" response yet, so by accepting a transaction it means it's "accepted and revised".

I've updated the text in https://github.com/stellar/go/pull/3642/commits/e3930041d4f73df61a3cc731a0a7613320a97c8d to reflect that. Please let me know if that's clearer now.

#### Migration files

> This project builds the migrations into a binary and embeds it into the built

Reviewer: 🔠 A binary makes it sound like the migrations end up in their own binary separate to the application. The binary is less confusing.
```suggestion
This project builds the migrations into the binary and embeds it into the built
```

Author: Fixed in https://github.com/stellar/go/pull/3642/commits/7d71fc1b58600aa86da5e134567f6635ea7a4c27.

      --network-passphrase string                      Network passphrase of the Stellar network transactions should be signed for (NETWORK_PASSPHRASE) (default "Test SDF Network ; September 2015")
      --port int                                       Port to listen and serve on (PORT) (default 8000)
```

## Account Setup

In order to properly use this server for regulated assets, the account whose
secret was added in `--issuer-account-secret (ISSUER_ACCOUNT_SECRET)` needs to
be configured according with SEP-8 [authorization flags] by setting both
`Authorization Required` and `Authorization Revocable` flags. This allows the
issuer to grant and revoke authorization to transact the asset at will.

You can use [this
link](https://laboratory.stellar.org/#txbuilder?params=eyJhdHRyaWJ1dGVzIjp7ImZlZSI6IjEwMCIsImJhc2VGZWUiOiIxMDAiLCJtaW5GZWUiOiIxMDAifSwiZmVlQnVtcEF0dHJpYnV0ZXMiOnsibWF4RmVlIjoiMTAwIn0sIm9wZXJhdGlvbnMiOlt7ImlkIjowLCJhdHRyaWJ1dGVzIjp7InNldEZsYWdzIjozfSwibmFtZSI6InNldE9wdGlvbnMifV19)

> to set those flags in Stellar Laboratory. Just click the link, fill the account

Reviewer: 🔠 You don't need the word Just, that can be dropped too.
```suggestion
to set those flags in Stellar Laboratory. Click the link, fill the account
```

Author: Fixed in https://github.com/stellar/go/pull/3642/commits/7d71fc1b58600aa86da5e134567f6635ea7a4c27.

address and sequence number, then the account secret and submit the transaction.

> After setting up the issuer account you can send some unities of the regulkated
> asset to a stellar account using our internal
> `friendbot/?addr={stellar_address}` endpoint. This endpoint is not part of the
> official SEP-8 Approval Server spec, it's a debug feature to allow accounts to
> test sending transactions (payments with the issuer's regulated asset) to the
> server.

Reviewer: 🔠 Some word suggestions:
```suggestion
After setting up the issuer account you can send some amount of the regulated
asset to a stellar account using the servers friendbot
`friendbot/?addr={stellar_address}` endpoint. The friendbot endpoint is not part of the
SEP-8 Approval Server specification, it's a debug feature that allows accounts to
test sending transactions containing payments with the issuer's regulated asset, to the
server.
```

Author: Fixed in https://github.com/stellar/go/pull/3642/commits/7d71fc1b58600aa86da5e134567f6635ea7a4c27.

to do that in Stellar Laboratory.

## API Spec

### `POST /tx-approve`

This is the core [SEP-8] endpoint used to validate and process
approval/revision/rejection of regulated assets transactions. Note: The example
responses below have set their `base-url` env var configured to

> `"https://sep8-base-url.com"`.

Reviewer: 💡 I recommend using example.com wherever possible for example domain names. The domain exists for this purpose, and prevents us from inadvertently linking to a potentially malicious website if someone registers the domain you have here.
```suggestion
`"https://example.com"`.
```

Author: Updated in: https://github.com/stellar/go/pull/3642/commits/7dec261924707f2935e494d4f58623fd67e749bc.

There are also some occurrences of `sep8-server.test` in the tests but I'll replace them in the next PR.

This endpoint is used for the extra action after `/tx-approve`, as described in
the SEP-8 [Action Required] section.

> Currently an arbitrarily criteria is implemented: email addresses starting with
> "x" will have the KYC automatically denied while all other emails will be
> accepted.

Reviewer: 🔠 Typo:
```suggestion
Currently an arbitrary criteria is implemented: email addresses starting with
"x" will have the KYC automatically denied while all other emails will be
accepted.
```

Author: Fixed in https://github.com/stellar/go/pull/3642/commits/7d71fc1b58600aa86da5e134567f6635ea7a4c27.
